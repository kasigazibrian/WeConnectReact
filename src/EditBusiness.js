import React from 'react';
import axios from "axios";
import {toast} from "react-toastify";
import NavigationBar from "./NavigationBar";
import {   FormGroup, Container, Col, Input, Label } from 'reactstrap';
import {withRouter, Redirect } from 'react-router-dom'

class EditBusiness extends React.Component{
    constructor(props){
        super(props);

        this.state={
            is_authenticated: this.props.getAuth(),
            business_name: "",
            business_category: "",
            business_location: "",
            business_email: "",
            contact_number: "",
            business_description: "",
            business_id: "",
            updated_successfully: false
        }
    }

    componentWillMount() {
        if (localStorage.getItem('token') === null) {
            this.setState({is_authenticated: true});
            this.props.history.push('/login');
            toast.error("Please login", {position: toast.POSITION.BOTTOM_CENTER});

        }
        else {
            this.setState({is_authenticated: true});
            const  id  = this.props.match.params.business_id;
            axios.get(`http://localhost:5000/api/v2/businesses/${id}`)
                .then(response=> {
                    console.log(response.data);
                    this.setState({business_name: response.data.Businesses[0].business_name});
                    this.setState({business_category: response.data.Businesses[0].business_category});
                    this.setState({business_location: response.data.Businesses[0].business_location});
                    this.setState({contact_number: response.data.Businesses[0].contact_number});
                    this.setState({business_email: response.data.Businesses[0].business_email});
                    this.setState({business_description: response.data.Businesses[0].business_description});
                    this.setState({business_id: response.data.Businesses[0].business_id});
                })
                .catch(error =>{
                    toast.error(error.response.data.Message, {position: toast.POSITION.BOTTOM_CENTER});
                    this.props.history.push('/');

                });
        };
    }
    handleBusinessNameUpdate = event => {
        this.setState({business_name: event.target.value});
        console.log(event.target.value)

    };

    handleBusinessCategoryUpdate = event => {
        this.setState({ business_category: event.target.value });
        console.log(event.target.value)

    };
    handleBusinessLocationUpdate = e => {
        this.setState({business_location: e.target.value});
        console.log(e.target.value)

    };
    handleBusinessEmailUpdate = e => {
        this.setState({business_email: e.target.value});
        console.log(e.target.value)
    };

    handleBusinessDescriptionUpdate = e =>{
        this.setState({business_description: e.target.value});
        console.log(e.target.value)
    };

    handleContactNumberUpdate = e =>{
        this.setState({contact_number : e.target.value});
        console.log(e.target.value)
    };
    handleUpdateSubmit = event => {
        const id = this.props.match.params.business_id;
        event.preventDefault();
        const business = {
            business_name: this.state.business_name,
            business_category: this.state.business_category,
            business_location: this.state.business_location,
            business_email: this.state.business_email,
            contact_number: this.state.contact_number,
            business_description: this.state.business_description
        };
        axios.defaults.headers.common['access-token'] = localStorage.getItem('token');
        axios.put(`http://localhost:5000/api/v2/businesses/${id}`,
            JSON.stringify(business),
            {headers: {'Content-Type': 'application/json'}
            })
            .then(res => {
                toast.success(res.data.Message, {position: toast.POSITION.BOTTOM_CENTER});
                console.log(res.data);
                this.setState({updated_successfully: true})
            })
            .catch(error => {
                toast.error(error.response.data.Message, {position: toast.POSITION.BOTTOM_CENTER});
                console.log(error.response.data);
            })
    };
render(){

    if (this.state.updated_successfully){
        return(
            <Redirect to={`/businesses/${this.state.business_id}`}/>
        )
    }

    return(
    <div>
        <NavigationBar auth={this.state.is_authenticated}/>
        <Container>
            <form onSubmit={this.handleUpdateSubmit}>
                <FormGroup>
                    <Col sm={"6"}>
                        <h1 style={{fontSize: "28px"}}>Edit Business </h1>
                    </Col>
                </FormGroup>
                <FormGroup>
                    <Col sm={"6"}>
                        <Label id="user">Business Name:</Label>
                        <input value={this.state.business_name} onChange={this.handleBusinessNameUpdate} type="text" style={{borderRadius: "20px"}} className="form-control" required="true"/>
                    </Col>
                </FormGroup>
                <FormGroup>
                    <Col sm={"6"}>
                        <Label id="user">Business Location:</Label>
                        <input value={this.state.business_location} onChange={this.handleBusinessLocationUpdate} type="text" style={{borderRadius: "20px"}} className="form-control" required="true"/>
                    </Col>
                </FormGroup>
                <FormGroup>
                    <Col sm={"6"}>
                        <Label id="user">Business Email:</Label>
                        <input value={this.state.business_email} onChange={this.handleBusinessEmailUpdate} type="text" style={{borderRadius: "20px"}} className="form-control" required="true"/>
                    </Col>
                </FormGroup>
                <FormGroup>
                    <Col sm={"6"}>
                        <Label for="contact_number">Contact Number:</Label>
                        <input value={this.state.contact_number} onChange={this.handleContactNumberUpdate} type="text" id="contact_number" style={{borderRadius: "20px"}}
                               className="form-control" required="true"/>
                    </Col>
                </FormGroup>
                <FormGroup>
                    <Col sm={"6"}>
                        <Label>Business Description:</Label>
                        <Input onChange={this.handleBusinessDescriptionUpdate} value={this.state.business_description} type="textarea" id="description" style={{borderRadius: "20px"}}
                               className="form-control" required="true"/>
                    </Col>
                </FormGroup>
                <FormGroup>
                    <Col sm={"6"}>
                        <Label for="exampleSelect">Business Category:</Label>
                        <Input onChange={this.handleBusinessCategoryUpdate} type="select" name="select" id="exampleSelect">
                            <option value="Entertainment">Entertainment</option>
                            <option value="Real Estate">Real Estate</option>
                            <option value="Education">Education</option>
                            <option value="Automobiles">Automobiles</option>
                            <option value="Health and Medicine">Health and Medicine</option>
                            <option value="Computers & Electronics">Computers & Electronics</option>
                        </Input>
                    </Col>
                </FormGroup>
                <FormGroup>
                    <Col sm={"6"}>
                        <button style={{borderRadius: "20px"}}
                                className={"btn btn-lg btn-info btn-block"}>Save Changes
                        </button>
                    </Col>
                </FormGroup>
            </form>
        </Container>
    </div>
    );
}
}
export default withRouter(EditBusiness)