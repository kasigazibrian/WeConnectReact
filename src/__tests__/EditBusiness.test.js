import React from 'react';
import { shallow, mount } from 'enzyme';
import EditBusiness from "../components/business/Editbusiness";
import MockAdapter from 'axios-mock-adapter';
import axios from "axios";
import Config from '../App.config'
import { MemoryRouter } from 'react-router-dom';


global.getAuth = ()=>{
    return false
}
global.match = {path: '/', url: "/edit_business/:business_id", params: {business_id: 1}, isExact: true }
let store = {};
window.localStorage = {
    getItem: key =>{return {"Token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VybmFtZSI6ImJyaWFuIiwiZXhwIjoxNTI5Njg1NTc3fQ.WJ2_sTwagTSBJ73iuBogIMVA6M8752ZUlCPOORNuWCI"}},
    setItem: (key, value)=> { store[key] = value},
    removeItem: key => Reflect.deleteProperty(store, key)
}
describe('Edit Business Component', ()=>{
    const wrapper = mount(<MemoryRouter><EditBusiness match={match}/></MemoryRouter>);
    const mock = new MockAdapter(axios);

    mock.onPut(`${Config.API_BASE_URL}/api/v2/businesses/1`).reply(201,
        {
            Message: "Business has been updated successfully",
            Status: "Success"
        })
        mock.onPut(`${Config.API_BASE_URL}/api/v2/businesses/undefined`).reply(400,
        {
            Message: "Business with this ID not found",
            Status: "Fail"
        })
        mock.onGet(`${Config.API_BASE_URL}/api/v2/businesses/1`).reply(200,
            {
                page: 1,
                limit: 20,
                count: 2,
                number_of_pages: 1,
                previous: null,
                next: null,
                Businesses: [
                  {
                    business_id: 1,
                    business_owner_id: 1,
                    business_name: "Mediacom",
                    business_email: "mediacom@gmail.com",
                    business_location: "Kampala",
                    contact_number: "256781712929",
                    business_category: "Real Estate",
                    business_description: "This business provides the best video coverage"
                  }
            ],
                Status: "Success"
        })
    it('it matches snapshot', () => {
        const wrapper = shallow(<EditBusiness match={match}/>)
        expect(wrapper.find(EditBusiness)).toMatchSnapshot();
    })
    
    it("Test it renders correctly", ()=>{
        expect(wrapper).toHaveLength(1)
        const component = wrapper.find(EditBusiness)
        expect(component).toHaveLength(1)

    });

    it('Should have a heading',()=>{
        const component = wrapper.find(EditBusiness)
        expect(component.find('h1#heading').text()).toContain("Edit Business")
    })
     
    it('Should have a save changes button',()=>{
        expect(wrapper.find('button.btn').text()).toContain("Save Changes")
    })
   
    it('check it changes state and value on input change', ()=> {
        const component = wrapper.find(EditBusiness)
        let businessName = component.find('input[name="businessName"]')
        businessName.simulate('change', {target: {name: "businessName",value: 'Supercom Limited'}});
        
        // wrapper.find('form')
        let businessLocation = component.find('input[name="businessLocation"]')
        businessLocation.simulate('change', {target: {name: "businessLocation",value: 'Kampala'}});

        let businessEmail = component.find('input[name="businessEmail"]')
        businessEmail.simulate('change', {target: {name: "businessEmail",value: 'supercom@gmail.com'}});
       
        let contactNumber = component.find('input[name="contactNumber"]')
        contactNumber.simulate('change', {target: {name: "contactNumber",value: '256781712927'}});

        let businessDescription = component.find('textarea[name="businessDescription"]')
        businessDescription.simulate('change', {target: {name: "businessDescription", value: 'This business provides the best services'}});

       
        let businessCategory = wrapper.find('select[name="businessCategory"]')
        businessCategory.simulate('change', {target: {name: "businessCategory", value: 'Entertainment'}});
       
        expect(businessName.instance().value).toBe('Supercom Limited');  
        expect(businessLocation.instance().value).toBe('Kampala');
        expect(businessEmail.instance().value).toBe('supercom@gmail.com');
        expect(contactNumber.instance().value).toBe('256781712927');
        expect(businessDescription.instance().value).toBe('This business provides the best services');
        expect(businessCategory.instance().value).toBe('Entertainment');
    });

    it('It handles errors on submit', async ()=>{
        const component = wrapper.find(EditBusiness)
        let store = {};
        window.localStorage = {
                getItem: key =>{return {"Token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VybmFtZSI6ImJyaWFuIiwiZXhwIjoxNTI5Njg1NTc3fQ.WJ2_sTwagTSBJ73iuBogIMVA6M8752ZUlCPOORNuWCI"}},
                setItem: (key, value)=> { store[key] = value},
                removeItem: key => Reflect.deleteProperty(store, key)
        }
        let businessName = component.find('input[name="businessName"]')
        businessName.simulate('change', {target: {name: "businessName",value: 'Supercom Limited'}});
        // wrapper.find('form')
        
        const form = wrapper.find('form')
        await form.simulate('submit',  { preventDefault: ()=>{}})
        
    });
    it('It handles submit', async ()=>{
        const EditBusinessComponent = mount(<EditBusiness match={match}/>)
        // console.log(EditBusinessComponent.props())
        let store = {};
        window.localStorage = {
                getItem: key =>{return {"Token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VybmFtZSI6ImJyaWFuIiwiZXhwIjoxNTI5Njg1NTc3fQ.WJ2_sTwagTSBJ73iuBogIMVA6M8752ZUlCPOORNuWCI"}},
                setItem: (key, value)=> { store[key] = value},
                removeItem: key => Reflect.deleteProperty(store, key)
        }
        let businessName = EditBusinessComponent.find('input[name="businessName"]')
        // businessName.simulate('change', {target: {name: "businessName",value: 'Supercom Limited'}});
        const form = wrapper.find('form')
        await form.simulate('submit',  { preventDefault: ()=>{}})
        
    });
    it('Should handle network error',()=>{
        mock.onPut(`${Config.API_BASE_URL}/api/v2/businesses/1`).networkError()
        const wrapper = mount(<EditBusiness match={match}/>)
        let spyOnComponentDidMount = jest.spyOn(wrapper.instance(), 'componentWillMount')
        // console.log(wrapper.instance())
        wrapper.instance().componentWillMount()
        wrapper.instance().handleUpdateSubmit({ preventDefault: ()=>{}})
        expect(spyOnComponentDidMount).toHaveBeenCalled()
    })
    it('Should handle bad request error',()=>{
        mock.onPut(`${Config.API_BASE_URL}/api/v2/businesses/1`).reply(400, {
            Message: "Not enough priviledge to perform action",
            Status: "Fail"
        })
        const wrapper = mount(<EditBusiness match={match}/>)
        let spyOnComponentDidMount = jest.spyOn(wrapper.instance(), 'componentWillMount')
        // console.log(wrapper.instance())
        wrapper.instance().componentWillMount()
        wrapper.instance().handleUpdateSubmit({ preventDefault: ()=>{}})
        expect(spyOnComponentDidMount).toHaveBeenCalled()
    })

    it('should set isAuthenticated state to false if token is null',()=>{
        let store = {};
      window.localStorage = {
          getItem: key =>{ return null},
          setItem: (key, value)=> { store[key] = value},
          removeItem: key => Reflect.deleteProperty(store, key)
          
      }
      const wrapper = shallow(<EditBusiness history={{push: ()=>{}}} match={match}/>)
      expect(wrapper.state('isAuthenticated')).toBe(false);  
    })
   
   
});