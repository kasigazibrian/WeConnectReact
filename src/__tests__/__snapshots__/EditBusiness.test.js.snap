// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Business Component it matches snapshot 1`] = `
ShallowWrapper {
  "length": 0,
  Symbol(enzyme.__root__): ShallowWrapper {
    "length": 1,
    Symbol(enzyme.__root__): [Circular],
    Symbol(enzyme.__unrendered__): <withRouter(EditBusiness) />,
    Symbol(enzyme.__renderer__): Object {
      "batchedUpdates": [Function],
      "getNode": [Function],
      "render": [Function],
      "simulateEvent": [Function],
      "unmount": [Function],
    },
    Symbol(enzyme.__node__): Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
    Symbol(enzyme.__nodes__): Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": [Function],
        },
        "ref": null,
        "rendered": [Function],
        "type": [Function],
      },
    ],
    Symbol(enzyme.__options__): Object {
      "adapter": ReactSixteenAdapter {
        "options": Object {
          "enableComponentDidUpdateOnSetState": true,
        },
      },
    },
  },
  Symbol(enzyme.__unrendered__): null,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
